File: src/css/properties/effects.zig
pub const css = @import("../css_parser.zig");

const SmallList = css.SmallList;

const CssColor = css.css_values.color.CssColor;
const Length = css.css_values.length.LengthValue;
const NumberOrPercentage = css.css_values.percentage.NumberOrPercentage;
const Angle = css.css_values.angle.Angle;
const Url = css.css_values.url.Url;

/// A value for the [filter](https://drafts.fxtf.org/filter-effects-1/#FilterProperty) and
/// [backdrop-filter](https://drafts.fxtf.org/filter-effects-2/#BackdropFilterProperty) properties.
pub const FilterList = union(enum) {
    /// The `none` keyword.
    none,
    /// A list of filter functions.
    filters: SmallList(Filter, 1),
};

/// A [filter](https://drafts.fxtf.org/filter-effects-1/#filter-functions) function.
pub const Filter = union(enum) {
    /// A `blur()` filter.
    blur: Length,
    /// A `brightness()` filter.
    brightness: NumberOrPercentage,
    /// A `contrast()` filter.
    contrast: NumberOrPercentage,
    /// A `grayscale()` filter.
    grayscale: NumberOrPercentage,
    /// A `hue-rotate()` filter.
    hue_rotate: Angle,
    /// An `invert()` filter.
    invert: NumberOrPercentage,
    /// An `opacity()` filter.
    opacity: NumberOrPercentage,
    /// A `saturate()` filter.
    saturate: NumberOrPercentage,
    /// A `sepia()` filter.
    sepia: NumberOrPercentage,
    /// A `drop-shadow()` filter.
    drop_shadow: DropShadow,
    /// A `url()` reference to an SVG filter.
    url: Url,
};

/// A [`drop-shadow()`](https://drafts.fxtf.org/filter-effects-1/#funcdef-filter-drop-shadow) filter function.
pub const DropShadow = struct {
    /// The color of the drop shadow.
    color: CssColor,
    /// The x offset of the drop shadow.
    x_offset: Length,
    /// The y offset of the drop shadow.
    y_offset: Length,
    /// The blur radius of the drop shadow.
    blur: Length,
};

File: src/css/properties/effects.rs
use crate::css_parser as css;

use css::SmallList;

use css::css_values::color::CssColor;
use css::css_values::length::LengthValue as Length;
use css::css_values::percentage::NumberOrPercentage;
use css::css_values::angle::Angle;
use css::css_values::url::Url;

/// A value for the [filter](https://drafts.fxtf.org/filter-effects-1/#FilterProperty) and
/// [backdrop-filter](https://drafts.fxtf.org/filter-effects-2/#BackdropFilterProperty) properties.
pub enum FilterList {
    /// The `none` keyword.
    None,
    /// A list of filter functions.
    Filters(SmallList<Filter, 1>),
}

/// A [filter](https://drafts.fxtf.org/filter-effects-1/#filter-functions) function.
pub enum Filter {
    /// A `blur()` filter.
    Blur(Length),
    /// A `brightness()` filter.
    Brightness(NumberOrPercentage),
    /// A `contrast()` filter.
    Contrast(NumberOrPercentage),
    /// A `grayscale()` filter.
    Grayscale(NumberOrPercentage),
    /// A `hue-rotate()` filter.
    HueRotate(Angle),
    /// An `invert()` filter.
    Invert(NumberOrPercentage),
    /// An `opacity()` filter.
    Opacity(NumberOrPercentage),
    /// A `saturate()` filter.
    Saturate(NumberOrPercentage),
    /// A `sepia()` filter.
    Sepia(NumberOrPercentage),
    /// A `drop-shadow()` filter.
    DropShadow(DropShadow),
    /// A `url()` reference to an SVG filter.
    Url(Url),
}

/// A [`drop-shadow()`](https://drafts.fxtf.org/filter-effects-1/#funcdef-filter-drop-shadow) filter function.
pub struct DropShadow {
    /// The color of the drop shadow.
    pub color: CssColor,
    /// The x offset of the drop shadow.
    pub x_offset: Length,
    /// The y offset of the drop shadow.
    pub y_offset: Length,
    /// The blur radius of the drop shadow.
    pub blur: Length,
}

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/css/properties/effects.zig (56 lines)
//   confidence: high
//   todos:      0
//   notes:      type-only definitions; SmallList<T, N> assumed in crate::css_parser
// ──────────────────────────────────────────────────────────────────────────
